Data/Vector.hs

```diff
@@ -219,11 +224,24 @@ data Vector a = Vector {-# UNPACK #-} !Int
                        {-# UNPACK #-} !(Array a)
         deriving ( Typeable )
 
+#if __GLASGOW_HASKELL__ >= 821
 instance NFData a => NFData (Vector a) where
-    rnf (Vector i n arr) = rnfAll i
+  rnf = liftRnf rnf
+  {-# INLINEABLE rnf #-}
+
+instance NFData1 Vector where
+  liftRnf elemRnf = foldl' (\_ -> elemRnf) ()
+  {-# INLINEABLE liftRnf #-}
+#endif
+
+#if __GLASGOW_HASKELL < 821
+instance NFData a => NFData (Vector a) where
```

Comment:
Having two entirely separate code paths here is a little unfortunate. I would prefer to abstract out the common bits into a top-level function (which isn't exported), which we might name:

```haskell
liftRnfVector :: (a -> ()) -> Vector a -> ()
liftRnfVector = ...
```

Once this is defined, we can implement the instances like so:

```haskell
instance NFData a => NFData (Vector a) where
  rnf = liftRnf rnf

#if MIN_VERSION_deepseq(1,4,3)
instance NFData1 Vector where
  liftRnf = liftRnfVector
#endif
```

This ensures that we use the same code paths everywhere, and it keeps CPP to a minimum.

Reply:
No problem.